presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java

```diff
@@ -0,0 +1,297 @@
+    }
+
+    @Override
+    public void addStateChangeListener(StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0.0;
+    }
+
+    @Override
+    public SharedBufferInfo getInfo()
+    {
+        OutputBuffer outputBuffer;
+        synchronized (this) {
+            outputBuffer = delegate;
+        }
+
+        if (outputBuffer == null) {
+            //
+            // NOTE: this code must be lock free to not hanging state machine updates
+            //
+            BufferState state = this.state.get();
+
+            return new SharedBufferInfo(
+                    state,
+                    state.canAddBuffers(),
+                    state.canAddPages(),
+                    0,
+                    0,
+                    0,
+                    0,
+                    ImmutableList.of());
+        }
+        return outputBuffer.getInfo();
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        Set<OutputBufferId> abortedBuffers = ImmutableSet.of();
+        List<PendingRead> pendingReads = ImmutableList.of();
+        OutputBuffer outputBuffer;
+        synchronized (this) {
+            if (delegate == null) {
+                delegate = new SharedBuffer(taskInstanceId, state, maxBufferSize, systemMemoryUsageListener);
+
+                // process pending aborts and reads outside of synchronized lock
+                abortedBuffers = ImmutableSet.copyOf(this.abortedBuffers);
+                this.abortedBuffers.clear();
+                pendingReads = ImmutableList.copyOf(this.pendingReads);
+                this.pendingReads.clear();
+            }
+            outputBuffer = delegate;
+        }
+
+        outputBuffer.setOutputBuffers(newOutputBuffers);
+
+        // process pending aborts and reads outside of synchronized lock
+        abortedBuffers.forEach(outputBuffer::abort);
+        for (PendingRead pendingRead : pendingReads) {
+            pendingRead.process(outputBuffer);
+        }
+    }
+
+    @Override
+    public CompletableFuture<BufferResult> get(OutputBufferId bufferId, long startingSequenceId, DataSize maxSize)
+    {
+        OutputBuffer outputBuffer;
+        synchronized (this) {
+            if (delegate == null) {
+                PendingRead pendingRead = new PendingRead(bufferId, startingSequenceId, maxSize);
+                pendingReads.add(pendingRead);
+                return pendingRead.getFutureResult();
+            }
+            outputBuffer = delegate;
+        }
+        return outputBuffer.get(bufferId, startingSequenceId, maxSize);
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
```

Comment:
Question:

* How do I search for usage a REST endpoint inside Presto?
* I thought `abort` is used to ack that the last page is received. I suppose I'm wrong. Otherwise, `abort` can't be called before `delegate` is set.